### DISCUSSION TITLE
How to show available objects to import using Pyright

### DISCUSSION THREAD

**Discussion by dbatten5** [hrsh7th/nvim-cmp#1785] [Q&A]
I have `pyright` set up which seems to be working fine as when I type in e.g. `from func`, an autocomplete suggestion of `functools` is displayed, see screenshot
<img width="330" alt="image" src="https://github.com/hrsh7th/nvim-cmp/assets/17271241/46d4d911-2c95-4aaf-8951-54eaf8a33116">
However, when accept the suggestion and write import, no autocomplete suggestions are displayed:
<img width="313" alt="image" src="https://github.com/hrsh7th/nvim-cmp/assets/17271241/0e9b75f9-4082-4fa3-99ae-afca7f27ca8e">
When I start typing a function I know exists in `functool`, e.g. `reduce`, then the autocomplete menu pops up
<img width="402" alt="image" src="https://github.com/hrsh7th/nvim-cmp/assets/17271241/47d3821b-a828-4dc2-a062-8ec9be42e906">
How can I set `nvim-cmp` up to automatically display the popup with available objects to import after I've written `from {x} import `?

**Comment by fitrh** [comment]
I think you can simply call `cmp.complete()`, which the insert preset binds to `<C-n>`/`<C-p>`, to show the completion items without typing any characters

**Reply by dbatten5** [reply]
Pressing `<c-p>` brings up this list
<img width="390" alt="image" src="https://github.com/hrsh7th/nvim-cmp/assets/17271241/94c4c3d9-7546-4564-bcf0-5049494862e3">
Looks like it's populating it with any symbol it can find in the current file

**Reply by fitrh** [reply]
Are you sure pyright is attached to the buffer? Because I just installed pyright to test it and this is what I got

<img width="1440" alt="Screenshot 2023-12-25 at 17 54 47" src="https://github.com/hrsh7th/nvim-cmp/assets/55179750/73472fca-9276-4591-9385-3057b356586b">

  
With the following LSP config

```lua
{
  settings = {
    python = {
      analysis = {
        stubPath = os.getenv("PYTHON_TYPE_STUBS") or "",
        typeCheckingMode = "basic",
        useLibraryCodeForTypes = true,
      },
      venvPath = ".",
    },
  },
}
```

**Reply by dbatten5** [reply]
Yeah it's definitely attached to the buffer as I get autocomplete suggestions when I start typing
<img width="426" alt="image" src="https://github.com/hrsh7th/nvim-cmp/assets/17271241/5c6248ce-17c0-45ef-ab03-c011879be0ef">
It's just I don't get them straight away, I need to start typing something. I figured it was something with the way I've set up `nvm-cmp`, maybe some trigger was missing.

I have the same LSP config as you
```
  settings = {
    python = {
      analysis = {
        stubPath = os.getenv("PYTHON_TYPE_STUBS") or "",
        typeCheckingMode = "basic",
        useLibraryCodeForTypes = true,
      },
    },
  },
```

**Reply by fitrh** [reply]
Are you sure your `<C-n>`/`<C-p>` is mapped to `cmp.complete()`?

**Reply by dbatten5** [reply]
This is the config I have set up currently
```
  {
      completion = {
        completeopt = "menu,menuone,preview,noselect",
      },
      window = {
        completion = cmp.config.window.bordered(),
        documentation = cmp.config.window.bordered(),
      },
      view = {
        docs = {
          auto_open = false,
        },
      },
      experimental = {
        ghost_text = true,
      },
      mapping = cmp.mapping.preset.insert({
        ["<c-p>"] = cmp.mapping.select_next_item(),
        ["<c-n>"] = cmp.mapping.select_prev_item(),
        ["<c-b>"] = cmp.mapping.scroll_docs(-4),
        ["<c-f>"] = cmp.mapping.scroll_docs(4),
        ["<c-g>"] = function()
          if cmp.visible_docs() then
            cmp.close_docs()
          else
            cmp.open_docs()
          end
        end,
        ["<c-space>"] = cmp.mapping.complete(),
        ["<c-e>"] = cmp.mapping.abort(),
        ["<cr>"] = cmp.mapping.confirm({ select = false }),
      }),
      sources = cmp.config.sources({
        {
          name = "nvim_lsp",
          entry_filter = function(entry, _)
            return require("cmp.types").lsp.CompletionItemKind[entry:get_kind()] ~= "Text"
          end,
        },
        { name = "nvim_lua" },
        { name = "buffer" },
        { name = "path" },
      }),
    }
```

If I add 
```
vim.keymap.set("i", "<c-p>", "<cmd>lua require('cmp').complete()<cr>")
```
to my config then I can open the autocomplete menu using `<c-p>`, which is alright, but what I really want is for it to open automatically

**Reply by fitrh** [reply]
Actually if you use `cmp.mapping.preset.insert` as your `mapping`, you don't need to define `<C-n>`/`<C-p>` since the preset alredy does that for you

```diff
      mapping = cmp.mapping.preset.insert({
-       ["<c-p>"] = cmp.mapping.select_next_item(),
-       ["<c-n>"] = cmp.mapping.select_prev_item(),
```

**Reply by fitrh** [reply]
>  but what I really want is for it to open automatically

Then you can define an autocmd like this

```lua
vim.api.nvim_create_autocmd(
  {"TextChangedI", "TextChangedP"},
  {
    callback = function()
      local line = vim.api.nvim_get_current_line()
      local cursor = vim.api.nvim_win_get_cursor(0)[2]

      local current = string.sub(line, cursor, cursor + 1)
      if current == "." or current == "," or current == " " then
        require('cmp').close()
      end

      local before_line = string.sub(line, 1, cursor + 1)
      local after_line = string.sub(line, cursor + 1, -1)
      if not string.match(before_line, '^%s+$') then
        if after_line == "" or string.match(before_line, " $") or string.match(before_line, "%.$") then
          require('cmp').complete()
        end
      end
  end,
  pattern = "*"
})
```

Though I found that behavaior too aggressive for someone like me who prefers manual trigger
